### ISSUE TITLE
Canonical enclave config format

### ISSUE THREAD

**Issue by wintersteiger** [lsds/sgx-lkl#327]
@letmaik previously summarized:

Dynamic attestation relies on extra data being passed to the enclave start function which then becomes part of the final MRENCLAVE hash and can be used in attestation. In OE terms this will be called EEID and is worked on by @wintersteiger. For SGX-LKL, the extra data is a concrete canonical serialization of the app config JSON so that the final MRENCLAVE can be predicted in advance and used in multi-party attestation. In the future, predicting the final MRENCLAVE will be unnecessary once attestation services support EEID natively and expose the JSON data as individual claims that can then be used in other services -- at that point, a canonical format is likely not needed anymore, but it may still be, due to extra restrictions (e.g. only allowing certain JSON structures).

This issue is about discussing the concrete canonical format.

Current proposal:

Format is still valid JSON.
All whitespace is removed.
All fields are sorted alphabetically.
Duplicates are not allowed in the input JSON.

This is a JSON without whitespaces and with keys sorted: `data_canonical = json.dumps(data, separators=(',', ':'), sort_keys=True)`

Extension that forbids duplicate keys in the input:

```
import json
def dict_raise_on_duplicates(ordered_pairs):
    # https://stackoverflow.com/a/14902564
    d = {}
    for k, v in ordered_pairs:
        if k in d:
           raise ValueError("duplicate key: %r" % (k,))
        else:
           d[k] = v
    return d

noncanonical = '{"b": 1, "a": 3}'
parsed = json.loads(noncanonical, object_pairs_hook=dict_raise_on_duplicates)
canonical = json.dumps(parsed , separators=(',', ':'), sort_keys=True)
# output: {"a":3,"b":1}
```

CC @prp @mikbras

**Comment by wintersteiger** [comment]
Further: JSON does not support integers greater than 2^53, so many integers will have to be encoded into strings. I've started doing that consistently by hex-encoding all integers into strings. As a side-effect this also bounds the lengths of those strings. I'll likely go all the way and add leading zeros so that all 32/64 bit integers are encoded into exactly 8/16 characters.

**Comment by letmaik** [comment]
We decided offline that all integers will be converted to a single decimal format wrapped in a string (`"1234"`), which allows to express numbers greater than 2^53 and keeps human-readability (compared to hex encoding). Small numbers are string-wrapped as well to make in-enclave parsing simpler (otherwise non-string integers have to be separately parsed and checked for out-of-range-ness).
One aspect I hadn't considered is how this will affect expressing rules for key release policies, since strings wouldn't allow numeric comparison anymore. The question is whether numeric comparisons (apart from identity) would ever be needed in such policies for any app config properties.

**Comment by wintersteiger** [comment]
That's right, key release policies would have to know about that encoding. In a pinch, strings are also numerically comparable (lexicographically), which, for instance, would work fine for version numbers.

**Comment by prp** [comment]
> That's right, key release policies would have to know about that encoding. In a pinch, strings are also numerically comparable (lexicographically), which, for instance, would work fine for version numbers.

I think that starting to compare numbers lexicographically or not supporting number comparisons in policies is going to be painful and lead to difficult to detect problems. A limitation of the format (JSON) shouldn't lead to broken policies.

How many app_config fields do we have that need numbers larger than 2^53? Could we just treat those in a special fashion e.g. by not using JSON's numeric type for them?

**Comment by wintersteiger** [comment]
That would complicate the parser and serializer code, making them harder to review for security and to predict for attestation purposes. I definitely don't want to treat some settings different from the others, that's bad even for user-readability. Note that this is for the (new) attested enclave config, not the json encoding for `sgxlkl_app_config_t`, which can stay user-readable.

I don't know how many we actually need, because it's not clear to me what they are used for, but even if we remove all the current ones, there'll be a new one tomorrow.

**Comment by prp** [comment]
I didn't mean to change the JSON parser but, as part of the schema, we would say that particular fields in the app_config, which we know may be large, must be strings. For example:
```
{
  "security_version" : 42,
  "key_nonce" : "473274979374937493",
...
}
```

**Comment by wintersteiger** [comment]
That requires a change to at least one of the JSON parsers, which now also needs a list of keys that are expect to be strings (or worse, either string or integer). Much safer to just encode all numbers as strings, which doesn't require any magic lists and the parser complexity is significantly lower and future-proof, and even user-readable if we keep them in decimal.

Also, it doesn't depend on one of us trawling through the code base to see whether any of these numbers could potentially be higher than 2^53 in any future version of the code, and then being wrong about it.

**Comment by prp** [comment]
I don’t understand. Doesn’t JSON parsing only parse values in quotes as strings?

**Comment by wintersteiger** [comment]
Yes, strings are quoted, but I don't understand your question. We have to check that they contain valid numbers (also in the parser), but it's the same for all numbers.

**Comment by prp** [comment]
Well, at some point, SGX-LKL specific code needs to interpret the parsed values and check ranges etc. This code will understand that some fields are strings and others are numbers, and it could translate some strings into numbers. It's not pretty, but the alternative is, as you say, to represent all numbers as strings, and I'm concerned how this will impact the policies that users write.

**Comment by wintersteiger** [comment]
Yes, that's what the JSON parser (or our callback therein) does, but it does it in the same way for all numbers. 

I don't think this would have a major impact on the policies that users would write. Do you expect that they would want to use many range checks, which they would be discouraged from if we use strings?

**Comment by prp** [comment]
Checking version/security numbers is the obvious one. In addition, there may be "safe" ranges for allocated memory amounts etc. Perhaps we should first come up with a full specification of an app_config and also think about what associated policies may look like?

**Comment by wintersteiger** [comment]
That's a very good idea indeed! I think it would be helpful to have such examples in discussions/meetings with others too, so it makes sense to actually go through a number of them and write them down.
I'm close to having a "full" specification (in the sense of all the settings that the code actually depends on), I hope to finish that tomorrow.

**Comment by wintersteiger** [comment]
This discussion has run a bit dry, we should think about this at some point and see whether we can make the config format truly canonical. 

Some notes from an out of band thread: I don't think we ever finished that discussion and just accepted the fact that JSON will never be truly canonical. For instance, strings would still allow leading zeros (although I think I disallowed them in the schema) and there may be an optional + sign. Also, the order of array elements is not necessarily fixed (I think it's fine in our case, but I'm not sure). I don't even want start thinking about Unicode (in the JSON standard). Then there are a bunch of issues with intentionally mis-constructed files. All these problems can probably be address in some way, but the right way to make sure is to prove it (e.g. let EverParse produce a verified parser for our subset of JSON) but nobody has time to do that at the moment.

**Comment by davidchisnall** [comment]
I would like to cleanly separate the provided config format from the EEID format.  The former should ideally be [UCL](https://github.com/vstakhov/libucl), the latter should be sorted, minified, JSON.  This would also address the desire to have comments, defaults, and so on in the configuration.  `ucl_object_sort_keys` can recursively sort all keys in a UCL document before generating compact JSON.

**Comment by wintersteiger** [comment]
There are other constraints than just sortedness and I don't think we should pull in an external library for the enclave side of this. On the host side it might make sense, but I think we gain so little that it's not worth the amount of effort and code duplication.

**Comment by davidchisnall** [comment]
You gain a huge amount from using libUCL on host side.  It will output compact JSON, so does not need to be present in the enclave.

**Comment by letmaik** [comment]
@davidchisnall I haven't heard of UCL before. I think what we really want is just JSON with comments. This would also work together with existing tooling like VS Code that can use the JSON schema for auto-completion etc. Having JSON with comments is a bonus though and shouldn't be a priority now.

@wintersteiger Your [notes](https://github.com/lsds/sgx-lkl/issues/327#issuecomment-650265020) are useful and I think should be made into a check list in the issue description, possibly referencing issues/PRs that fix those problems.
